libyul/AsmParser.cpp

```diff
@@ -567,18 +567,20 @@ std::variant<Literal, Identifier> Parser::parseLiteralOrIdentifier(bool _unlimit
 			break;
 		}
 
+		auto const beginOfLiteralLoc = currentLocation();
 		Literal literal{
 			createDebugData(),
 			kind,
-			valueOfLiteral(currentLiteral(), kind, _unlimitedLiteralArgument && kind == LiteralKind::String),
-			kind == LiteralKind::Boolean ? m_dialect.boolType : m_dialect.defaultType
+			valueOfLiteral(currentLiteral(), kind, _unlimitedLiteralArgument && kind == LiteralKind::String)
 		};
 		advance();
 		if (currentToken() == Token::Colon)
 		{
 			expectToken(Token::Colon);
 			updateLocationEndFrom(literal.debugData, currentLocation());
-			literal.type = expectAsmIdentifier();
+			auto const loc = SourceLocation::smallestCovering(beginOfLiteralLoc, currentLocation());
```

Comment:
It would be much more helpful if these names said what is the thing the location of which they hold.

```suggestion
			auto const typedLiteralLocation = SourceLocation::smallestCovering(literalLocation, currentLocation());
```

Same in `parseNameWithDebugData()`.

Reply:
By the way, I'd expect a function named `expectUntyped()` to basically include this whole `if`. If it's just reporting an error it would be clearer if its name said just that.

The `loc` argument in that function is also missing an underscore and should have a more self-explanatory name too (e.g. `_typedLiteralLocation`).

Reply:
I have renamed the function from `expectUntyped` to `raiseUnsupportedTypesError`. That should be more descriptive. As for its argument (aside from the underscore I failed to put) I think that keeping it more general doesn't hurt. You could, potentially, have also other sorts of locations which are associated to that error, so why artificially restrict it.

Reply:
> As for its argument (aside from the underscore I failed to put) I think that keeping it more general doesn't hurt. 

Fair enough. Looks like you forgot to adjust the name in the header though (https://github.com/ethereum/solidity/pull/15332#discussion_r1733127855).